README

                       The MLKit sources

This README file describes the structure of the MLKit sources as well
as necessary steps for compilation and either personal or system wide
installation.

An overview of the directory structure of the distribution is given in
Appendix A. 

*NOTE*: For a quick compilation and installation guide, see Appendix B.

1. Introduction
---------------
The MLKit is a compiler for the programming language Standard ML. The
MLKit covers all of Standard ML, as defined in the 1997 edition of the
Definition of Standard ML and supports most of the Standard ML Basis
Library.

Features:

  * All of Standard ML. The MLKit compiles all of Standard ML,
    including Modules, as specified by the Definition of Standard
    ML. The MLKit also supports large parts of the Standard ML Basis
    Library.

  * Supports ML Basis Files: The MLKit compiles large programs,
    including itself, around 80.000 lines of Standard ML plus the
    Standard ML Basis Library. The support for ML Basis Files
    makes it easy to compile a program with different Standard ML
    compilers. Currently, both MLton and the MLKit supports the
    concept of ML Basis Files.
 
  * Region-Based Memory Management: Memory allocation directives (both
    allocation and deallocation) are inferred by the compiler, which
    uses a number of program analyses concerning lifetimes and storage
    layout. The MLKit compiler is unique among ML implementations in
    this respect.

  * Reference-tracing Garbage Collection: The MLKit supports
    reference-tracing garbage collection in combination with region
    memory management.

  * Native backend for the x86 architecture.

  * Documentation. A comprehensive guide on programming with the MLKit
    is available in the doc/ directory. Documentation is also
    available in man-pages and from the MLKit home page

                http://melsman.github.io/mlkit

2. License and Copyright
------------------------
The MLKit compiler is distributed under the GNU Public License,
version 2. See the file doc/license/MLKit-LICENSE for details. The
runtime system (kit/src/Runtime/) and libraries (kit/basis/) is
distributed under the more liberal MIT License.

3. Compilation Requirements
---------------------------
To compile, install, and use the MLKit, a Linux box running Ubuntu
Linux, Debian, gentoo, or similar is needed. The MLKit also works on
Mac OS and has also earlier been reported to run on the FreeBSD/x86
platform, with a little tweaking.

To compile the MLKit, a Standard ML compiler is needed, which needs to
be one of the following:

 (1) MLton >= 20051202. A link to the home page for MLton is available
     from the MLKit home page. The command sml should be available
     from the prompt:

       $ mlton
       MLton 20051202 (built Sat Dec 03 04:20:11 2005 on pavilion)

 (2) A working MLKit compiler >= 4.3.0. The command mlkit should be
     available from the prompt:

       $ mlkit -V
       MLKit version 4.3.0, Jan 25, 2006 [X86 Backend]

Moreover, Gcc is needed for compiling the runtime system and related
tools.

4. Compilation
--------------
After having checked out the sources from Github, execute the command:
 
 $ ./autobuild

Now, cd to the toplevel directory of the package. Execute the commands
under (a) or (b):

  a. Compile with MLton alone (Tested with 3Gb RAM)
  -------------------------------------------------
   $ ./configure 
   $ make mlkit

  b. Compile with existing MLKit (Tested with 1Gb RAM)
  ----------------------------------------------------
   $ ./configure --with-compiler=mlkit
   $ make mlkit

If you later want to install the MLKit in your own home directory, you
should also pass the option --prefix=$HOME/mlkit to ./configure above.

For binary packages, we use

   $ ./configure --sysconfdir=/etc --prefix=/usr

5. Bootstrapping (optional - works with 1Gb RAM)
------------------------------------------------
This step is optional. If you want the resulting executable compiler
to be bootstrapped (compiled with itself), execute the command:

 $ make bootstrap

Be aware that this step takes some time.

6. Pre-compile Basis Library and Kit-Library
--------------------------------------------
Execute the following commands:

 $ make mlkit_libs

7. Installation
---------------
For a system wide installation of the MLKit, installation including
man-pages and tools, execute the commands - as root:

 % make install

For a personal installation, with --prefix=$HOME/mlkit given to
./configure, execute the following commands:

 $ make install

8. Making a Binary Package
--------------------------
To build a binary package, execute the command

 $ make mlkit_i386_tgz

This command leaves a package mlkit-X.Y.Z-i386.tgz in the dist/
directory. For building a binary package, step 7 is not needed and
step 5 is optional.

9. Try It
---------
To test the installation, copy the directory /usr/share/mlkit/kitdemo to
somewhere in your own directory, say $HOME/kitdemo:

 $ cp -a /usr/share/mlkit/kitdemo $HOME/kitdemo
 $ cd $HOME/kitdemo
 $ mlkit helloworld.sml

The MLKit should produce an executable file "run":  

 $ ./run
 hello world

10. More Information
--------------------
See the MLKit home page

   http://melsman.github.io/mlkit

Documentation for the MLKit is located in the directories doc/mlkit
and man/man1. License information is located in the file
doc/license/MLKit-LICENSE.

11. VCG
-------
The VCG tool, which is used to show region flow graphs, can be found
at http://www.cs.uni-sb.de/RW/users/sander/html/gsvcg1.html

12. Comments and Bug Reports
----------------------------
Please see the MLKit home page for a list of known bugs and
limitations. To file a bug-report, create an issue at the Github page.

Appendix A: Directory Structure of the Sources
----------------------------------------------

     kit/
        README
        configure
        Makefile.in        
        src/
        basis/
        doc/mlkit.pdf
           /license/MLKit-LICENSE                   
        man/man1/rp2ps.1                
        kitdemo/
        test/

Appendix B: Quick Compilation and Installation Guide
----------------------------------------------------
We assume that MLton >= 20051202 is installed on the system as
described in Section 3.

After having checked out the sources from Github, execute the command:
 
 $ ./autobuild

To compile and install the MLKit, execute the following commands:

 $ ./configure
 $ make mlkit
 $ make bootstrap
 $ make mlkit_libs

Execute as root:

 % make install

See Section 9 for how to test the installation.

NEWS
MLKit NEWS file

              **** MLKit version 4.3.9 is released ****

 * mael 2016-03-12: Support for parallel builds using 'make -j'.

 * mael 2016-03-12: Eliminated hard dependency on pdflatex.

 * mael 2016-03-12: Modified KitTester to generate HTML reports
   instead of pdf through LaTeX.

              **** MLKit version 4.3.8 is released ****

 * mael 2016-03-09: Fixed issue with assembler invocation on Mac OS X.

 * mael 2015-05-11: Fixed a few runtime system compiler warnings.

 * mael 2015-05-08: Fixed inefficiency in region inference algorithm
   for programs containing highly continuation-based nested recursive
   functions.

 * mael 2014-05-23: Moved the sources to github. Updated documentation
   and README files.

              **** MLKit version 4.3.7 is released ****

 * mael 2012-08-29: Port of the MLKit to Mac OS X. Still the backend
   generates 32-bit code.

 * mael 2012-08-29: Fixed bug in Posix.ProcEnv.uname.

 * mael 2012-04-24: Automatic enabling of aggressive optimisations
   when garbage collection is enabled.

              **** MLKit version 4.3.6 is released ****

 * mael 2012-04-23: Cleaning up source code and allowed for the
   compiler and runtime system to build on 64-bit systems, using
   the -m32 flag to gcc, gas, and ld. 

 * mael 2011-05-31: Renamed many files to allow for checkout of
   repository on case-insensitive file systems.

 * mael 2011-02-23: Optimized sizes of pickles generated for symbol
   table information. Now, using the new Bitstream module in
   src/Pickle, it is possible to output single bits to the
   stream. Also reverted pickler dictionaries for sharing to be
   localized for each type, which opens up for using short
   locators. Sizes are now approximately half the size of before.

 * mael 2011-02-23: Fixed Posix.c bug when compiling runtime system
   under ubuntu > 9.10 (reported by Niels Hallenberg).

 * mael 2008-10-06: Compiling with SML/NJ is no longer supported.

 * mael 2007-11-14: Fixed bug reported by Vesa Karvonen concerning
   unification of region types.

 * mael 2007-11-09: Improved type checking of the internal language
   LambdaExp. Type schemes are now normalized (and inforced to be so)
   so that bound names always occur in the body of the type scheme
   prior to region inference. This normalization happens after
   compilation into the LambdaExp language.

 * mael 2007-11-06: Improved type checking of the internal language
   LambdaExp. LambdaExp programs are now closed w.r.t. type variables.

 * mael 2007-11-02: Fixed bug in opacity elimination. Bug reported by
   Vesa Karvonen; see test/vesa2.sml for a small example that didn't work
   before the bug fix.

 * mael 2007-11-01: Support added for taking mlb-path-map-files as
   MLKit command-line arguments using the option
   --mlb_path_maps. Patch supplied by Vesa Karvonen.

 * mael 2007-10-31: Infinite precision integers are now allowed in
   patterns. See file test/intinf2.sml.

 * mael 2007-10-26: Signature BOOL now specifies true and
   false. Signature LIST specifies :: and nil. Structure List declares
   :: and nil. Structure Bool declares true and false.

 * mael 2007-10-26: Improved error message for a particular kind of
   overloading resolution error. Elaboration of the program

      val _ = fn (a,b) => (a + 1) + (b + 0w1)

   results in the introduction of a type variable with an empty set
   of overload possibilities. We now check during unification that
   overloading sets do not become empty...

 * mael 2007-10-26: Bug fix: array types should admit equality even if
   the argument to the array type constructor does not. Bug report
   submitted by Vesa Karvonen.

 * mael 2007-07-04: Error messages for errors in mlb-files have been
   improved; see test/error1.mlb, test/error2.mlb, test/error3.mlb,
   and test/error4.mlb.

 * mael 2007-04-19: Executing mlkit with no command args now returns
   minimal help info.

 * mael 2007-04-12: Functor bodies are now stored directly in module
   interpretation bases. Upon compilation of a functor application,
   the functor body is extracted from the interpretation basis and
   compiled. Previously, functor bodies were stored in designated
   functor body files, which led to problems with moving compilation
   bases around for package installations, etc.

 * mael 2007-03-22: Added mlkit-mlyacc executable and man-pages for
   mlkit-mlyacc and mlkit-mllex.

 * mael 2007-03-12: Added mlkit-mllex executable by including a
   modified copy of mllex from SML/NJ 110.60. The copy of the source
   code is in src/Tools/ml-lex. The source code is compiled with
   either mlton or with mlkit, depending on the setting of MLCOMP at
   configure-time.

 * mael 2006-10-29: Applied patch by Vesa Karvonen. The patch:

     - Exposes the Basis library Text : TEXT module, which wasn't
       exposed for some reason and also removes the opaque signature
       ascription which would (incorrectly) generate new types char,
       string, vector, ...

     - Extends the MLB parser to handle an arbitrary number of
       annotations in MLB files.  Unrecognized annotations result in a
       warning and have no other semantic effect.

     - Extends the MLB lexer to recognize simple quoted strings.  This
       is required to parse MLton's annotations properly.

     - Extends the expansion of MLB path variables so that path
       variables are allowed at any position of a path rather than
       just at the beginning.

   --Vesa Karvonen

              **** MLKit version 4.3.0 is released ****

 * mael 2006-01-27: Summary of major changes since 4.1.4

      - Updated Basis Library

      - Support for separate compilation with MLB-files

      - Garbage collection of regions is now the default

      - Improved documentation - man-pages, wiki, etc.

      - Various bug-fixes

 * mael 2006-01-27: Added Makefile entry for constructing a source
   distribution (in directory kit/dist/). Run 'make mlkit_src_tgz'.


js/index.php
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html>
<head>
<META http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>SMLtoJs - A Standard ML to JavaScript Compiler</title>
<link href="base.css" rel="stylesheet" type="text/css">
</head>
<body>
<table border=0 cellpadding=0 cellspacing=0 width=100%>
<tr><td width=75%>
<img src=smltojs_logo_color.png>
<h2>A Standard ML to JavaScript Compiler</h2>
<h4>By <a href="http://www.elsman.com">Martin Elsman</a>, 2008-08-16</h4>

<hr>

<h3>Introduction</h3>
<p>
SMLtoJs (pronounced "SML toys") is a compiler from Standard ML to
JavaScript, which allows programmers to enjoy the power of Standard ML
static typing, higher-order functions, pattern matching, and modules
for programming client-side web applications.
</p>

<p>SMLtoJs compiles all of Standard ML, including most of the Standard ML
Basis Library. It also has support for calling JavaScript functions
and for executing plain JavaScript.
</p>

<h3>Features</h3>

<ul>

<li><b>All of Standard ML</b>. SMLtoJs has support for all of Standard ML,
   including modules, pattern matching, higher-order functions,
   generative exceptions, etc.</li>

<li><b>Standard ML Basis Library support.</b> SMLtoJs has support for most of
   the Standard ML basis library, including the following structures:
      <code>Array2 ArraySlice Array Bool Byte Char CharArray CharArraySlice
      CharVector CharVectorSlice Date General Int Int31 Int32 IntInf
      LargeWord ListPair List Math Option OS.Path Pack32Big
      Pack32Little Random Real StringCvt String Substring Text Time
      Timer Vector VectorSlice Word Word31 Word32 Word8 Word8Array
      Word8ArraySlice Word8Vector Word8VectorSlice</code>.
</li>

<li><b>JavaScript integration</b>. SMLtoJs has support for calling JavaScript
   functions and for executing plain JavaScript. Moreover, Standard ML
   functions may be exported and called by JavaScript code.</li>

<li><b>Simple DOM access</b>. Simple DOM access and support for installing Standard ML
   functions as DOM event handlers and timer call back functions.</li>

<li><b>Optimization</b>. All Standard ML module language constructs, including functors,
   functor applications, and signature constraints, are eliminated by
   SMLtoJs at compile time. Moreover, SMLtoJs performs a series of
   compile time optimizations, including function inlining and
   specialization of higher-order recursive functions, such as map
   and foldl. Optimizations can be controlled using compile-time
   flags. As a result, SMLtoJs generates fairly efficient JavaScript
   code, although there are rooms for improvements; see below.</li>

<li><b>Reactive Web Programming</b>. SMLtoJs has library support for
Reactive Web Programming. See the <a href="test/rwp_ex1.html">rwp_ex1
example</a> for an introduction to what you can do with this
library. The <a href="test/rwp.sig">simple library API</a> captures
the basic notions of behaviors and event streams as well as the
concepts of behavior transformers and event stream transformers.</li>
</ul>

<h3>Online Demonstration</h3>

<p>
To see SMLtoJs in action, click on the examples in the table to the right; there are also
links to the Standard ML source code.
</p>

<h3>Getting the Sources</h3>
<p>
SMLtoJs compiles on Debian Linux systems with MLton or MLKit
installed. The SMLtoJS sources are available through an MLKit
sourceforge svn checkout:
<pre>
 $ svn co https://mlkit.svn.sourceforge.net/svnroot/mlkit/tags/smltojs-4.3.5/kit smltojs
</pre>
This command copies the sources to the directory smltojs.
</p>

<h3>Building SMLtoJs</h3>
<p>
To compile SMLtoJs from the sources (MLKit svn checkout;
see above), simply type
<pre>
  $ cd smltojs
  $ ./autobuild
  $ ./configure
  $ make smltojs
</pre>
If compilation succeeds, an executable
file <a href="smltojs.man.html"><code>bin/smltojs</code></a> should now be available.</p>

<h3>How it Works</h3>
<p>
The SMLtoJs executable <a href="smltojs.man.html"><code>bin/smltojs</code></a> takes as argument an sml-file
(or an mlb-file referencing the sml-files and other mlb-files of the
project) and produces an html file called <code>run.html</code> provided there are
no type errors! The resulting html-file mentions the generated
JavaScript files and a file <code>prims.js</code>, which contains a set of
primitive JavaScript functions used by the generated code.
</p>
<p>
Hint: Adding the flag <code>"-o name"</code> as command-line argument to <a href="smltojs.man.html"><code>smltojs</code></a>
results in the file name.html being generated instead of <code>run.html</code>.
</p>

<h3>Testing that it Works</h3>
<p>
To compile and test the test programs, cd to the <code>js/test</code> directory
and run <code>"make clean all"</code>:
<pre>
 $ cd js/test
 $ make clean all
</pre>
You can now start Firefox on the generated html-files; the file
all.html includes links to all the test files:
<pre>
 $ firefox all.html
</pre>
The examples temp.html, counter.html, and life.html are the most
interesting examples at the moment (more will come).
</p>

<h3>Issues</h3>
<p>
Currently, SMLtoJs does not implement tail-calls properly, as all
calls are translated into JavaScript calls. We are planning in the
near future to implement tail-calls using trampolines.
</p>
<p>
There is a known issue with a bug in the following test (in some cases, 
the implementation pretty prints reals slightly different than suggested by the spec):
<pre>
  real
  ----
   test13c: WRONG
</pre>
</p>
<p>
There are plenty of possibilities for further improvements, including:
<ol>

<li>Using numbers to represent constructors (instead of strings).</li>

<li>Unboxing of datatypes; currently, a boxed representation is used
   for lists, for example.</li>

</ol>
</p>

<h3>Copyright and License</h3>

<p>SMLtoJs is distributed under the GPL license; see the files in
<code>smltojs/doc/license</code> for copyright notices and detailed
license information.  
</p>

</td>

<td align=right valign=top width="25%">

   <table border=0 cellpadding=3 cellspacing=0 width=95% align=right>
    <tr><td><table border=0 bgcolor="#1f94b7" cellpadding=1 cellspacing=1 width=100% align=right>
         <tr><td><table border=0 bgcolor=white cellpadding=3 cellspacing=0 width=100%>
                   <tr><td bgcolor="#1f94b7"><font color=white><b>News</b></font></td></tr>
                   <tr><td bgcolor="#eeeeee"><font size=-1>
                       <b>2008-08-16: </b>SMLtoJs <b>version 4.3.5</b> is available.</td></tr>
                   <tr><td bgcolor="#eeeeee"><font size=-1>
                       <b>2007-11-27: </b>Martin gave a talk in the DIKU TOPPS group (<a href="slides_diku_2007-11-27.pdf">slides</a>).</td></tr>
                   <tr><td bgcolor="#eeeeee"><font size=-1>
                       <b>2007-09-17: </b>Martin gave a talk on the PLS Monday lunch meeting (<a href="slides_lunch_2007-09-17.pdf">slides</a>).</td></tr>
                   <tr><td bgcolor="#eeeeee"><font size=-1>
                       <b>2007-09-05: </b>SMLtoJs <b>version 4.3.4</b> is available.</td></tr>
                   <tr><td bgcolor="#eeeeee"><font size=-1>
                       <b>2007-08-31: </b>SMLtoJs <b>version 4.3.3</b> is available.</td></tr>
                   <tr><td bgcolor="#eeeeee"><font size=-1>
                       <b>2007-08-08: </b>SMLtoJs <b>version 4.3.2</b> is available.</td></tr>
                 </table>
             </td>
         </tr>
        </table></td></tr><tr><td>&nbsp;</td></tr><tr><td>
<table border=0 bgcolor="#1f94b7" cellpadding=1 cellspacing=1 width=100% align=right>
         <tr><td><table border=0 bgcolor=white cellpadding=3 cellspacing=0 width=100%>
                   <tr><td bgcolor="#1f94b7"><font color=white><b>SMLtoJs Examples</b></font></td></tr>
                   <tr><td bgcolor="#eeeeee"><table width=100%><tr><td><font size=-1><b>program</b></font></td><td>

           <font size=-1><b>sources</b></font></td></tr>
<tr><td>
<a href="test/rwp_ex1.html">Ex: rwp_ex1</a> </td><td><a href="test/rwp_ex1.mlb.html">src</a></td></tr><tr><td>
<a href="test/rwp_ex2.html">Ex: rwp_ex2</a> </td><td><a href="test/rwp_ex2.mlb.html">src</a></td></tr><tr><td>
<a href="test/rwp_ex3.html">Ex: rwp_ex3</a> </td><td><a href="test/rwp_ex3.mlb.html">src</a></td></tr><tr><td>
<a href="test/counter.html">Ex: counter</a> </td><td><a href="test/counter.mlb.html">src</a></td></tr><tr><td>
<a href="test/temp.html">Ex: temp</a> </td><td><a href="test/temp.mlb.html">src</a></td></tr><tr><td>
<a href="test/life.html">Ex: life</a> </td><td><a href="test/life.mlb.html">src</a></td></tr><tr><td>
<a href="test/cal.html">Ex: cal</a> </td><td><a href="test/cal.mlb.html">src</a></td></tr><tr><td>
<a href="test/array2.html">Test: array2</a> </td><td><a href="test/array2.sml">src</a></td></tr><tr><td>
<a href="test/int31.html">Test: int31</a> </td><td><a href="test/int31.sml">src</a></td></tr><tr><td>
<a href="test/int.html">Test: int</a> </td><td><a href="test/int.sml">src</a></td></tr><tr><td>
<a href="test/vectorslice.html">Test: vectorslice</a> </td><td><a href="test/vectorslice.sml">src</a></td></tr><tr><td>
<a href="test/word8arrayslice.html">Test: word8arrayslice</a> </td><td><a href="test/word8arrayslice.sml">src</a></td></tr><tr><td>
<a href="test/word8vector.html">Test: word8vector</a> </td><td><a href="test/word8vector.sml">src</a></td></tr><tr><td>
<a href="test/arrayslice.html">Test: arrayslice</a> </td><td><a href="test/arrayslice.sml">src</a></td></tr><tr><td>
<a href="test/general.html">Test: general</a> </td><td><a href="test/general.sml">src</a></td></tr><tr><td>
<a href="test/int32_2.html">Test: int32_2</a> </td><td><a href="test/int32_2.sml">src</a></td></tr><tr><td>
<a href="test/math.html">Test: math</a> </td><td><a href="test/math.sml">src</a></td></tr><tr><td>
<a href="test/stringcvt.html">Test: stringcvt</a> </td><td><a href="test/stringcvt.sml">src</a></td></tr><tr><td>
<a href="test/testdyn1.html">Test: testdyn1</a> </td><td><a href="test/testdyn1.sml">src</a></td></tr><tr><td>
<a href="test/vector.html">Test: vector</a> </td><td><a href="test/vector.sml">src</a></td></tr><tr><td>
<a href="test/word8array.html">Test: word8array</a> </td><td><a href="test/word8array.sml">src</a></td></tr><tr><td>
<a href="test/array.html">Test: array</a> </td><td><a href="test/array.sml">src</a></td></tr><tr><td>
<a href="test/int_2.html">Test: int_2</a> </td><td><a href="test/int_2.sml">src</a></td></tr><tr><td>
<a href="test/int32.html">Test: int32</a> </td><td><a href="test/int32.sml">src</a></td></tr><tr><td>
<a href="test/listpair.html">Test: listpair</a> </td><td><a href="test/listpair.sml">src</a></td></tr><tr><td>
<a href="test/string.html">Test: string</a> </td><td><a href="test/string.sml">src</a></td></tr><tr><td>
<a href="test/testmatc.html">Test: testmatc</a> </td><td><a href="test/testmatc.sml">src</a></td></tr><tr><td>
<a href="test/word.html">Test: word</a> </td><td><a href="test/word.sml">src</a></td></tr><tr><td>
<a href="test/word8.html">Test: word8</a> </td><td><a href="test/word8.sml">src</a></td></tr><tr><td>
<a href="test/word31.html">Test: word31</a> </td><td><a href="test/word31.sml">src</a></td></tr><tr><td>
<a href="test/word32.html">Test: word32</a> </td><td><a href="test/word32.sml">src</a></td></tr><tr><td>
<a href="test/bytechar.html">Test: bytechar</a> </td><td><a href="test/bytechar.sml">src</a></td></tr><tr><td>
<a href="test/int31_2.html">Test: int31_2</a> </td><td><a href="test/int31_2.sml">src</a></td></tr><tr><td>
<a href="test/intinf.html">Test: intinf</a> </td><td><a href="test/intinf.sml">src</a></td></tr><tr><td>
<a href="test/list.html">Test: list</a> </td><td><a href="test/list.sml">src</a></td></tr><tr><td>
<a href="test/real.html">Test: real</a> </td><td><a href="test/real.sml">src</a></td></tr><tr><td>
<a href="test/substring.html">Test: substring</a> </td><td><a href="test/substring.sml">src</a></td></tr><tr><td>
<a href="test/word8vectorslice.html">Test: word8vectorslice</a> </td><td><a href="test/word8vectorslice.sml">src</a></td></tr><tr><td>
<a href="test/time.html">Test: time</a> </td><td><a href="test/time.sml">src</a></td></tr><tr><td>
<a href="test/date.html">Test: date</a> </td><td><a href="test/date.sml">src</a></td></tr><tr><td>
<a href="test/unixpath.html">Test: unixpath</a> </td><td><a href="test/unixpath.sml">src</a></td></tr><tr><td>
<a href="test/exec.html">Test: exec</a> </td><td><a href="test/exec.mlb.html">src</a></td></tr><tr><td>
                 </table>
             </td>
         </tr>
        </table>

</td>
</tr>
</table>
</td></tr></table>
</td>
</tr>
</table>
<hr>
<a href="mailto:mael@elsman.com">mael@elsman.com</a>
</body>
</html>
